### ISSUE TITLE
Question about the LiLora aux matrix

### ISSUE THREAD

**Issue by 1585231086** [KohakuBlueleaf/HyperKohaku#16]
I noticed that the original paper stated that down_aux/up_aux is a progressive orthogonal vector, but the code in this repository orthogonalized the auxiliary layer weights with down_dim/out_feature_dim. This may be the reason why losses cannot be converged. Can you take a look at it? Thanks!

**Comment by KohakuBlueleaf** [comment]
@1585231086 
![image](https://github.com/KohakuBlueleaf/HyperKohaku/assets/59680068/6c49a660-bac1-47f7-bc00-1c6855fda841)

I checked the paper again
I think I'm doing it in the right way

I also contected with author to ensure that I can just use pytorch's orthogonal
if you have better method can you just try to modify the code and make a PR?

**Comment by chenxinhua** [comment]
> I noticed that the original paper stated that down_aux/up_aux is a progressive orthogonal vector, but the code in this repository orthogonalized the auxiliary layer weights with down_dim/out_feature_dim. This may be the reason why losses cannot be converged. Can you take a look at it? Thanks!

In the pre-optimization stage, the loss can converge normally, indicating that the implementation in Lilora is correct. However, after exporting the Lilora weights to LoRA, the inference testing results are distorted.

**Comment by chenxinhua** [comment]
The primary issue is to ensure that the training and exporting of Lilora work properly, Lilora consists of a state aux_seed and learnable parameters weights. I trained Lilora using a single image, and the loss function converged normally. However, after exporting it as Lora, it fails to generate accurate results.

**Comment by 1585231086** [comment]
> 


i think so, too

**Comment by 1585231086** [comment]
> The primary issue is to ensure that the training and exporting of Lilora work properly, Lilora consists of a state aux_seed and learnable parameters weights. I trained Lilora using a single image, and the loss function converged normally. However, after exporting it as Lora, it fails to generate correct results.


I think so, too

**Comment by garychan22** [comment]
When I tried to resume from the latest checkpoint for further training of the hypernetwork, the validation results suddenly become weird and far from the ones generated from the previous iterations, there may be some issues about saving/loading lilora (weights/rng_state)?

**Comment by chenxinhua** [comment]
Simply changing the down_aux and up_aux in Lilora from random orthogonal matrices to learnable matrices, and training it with a single image, the loss function converges normally. After exporting Lilora to regular Lora, the inference testing successfully generates correct facial images. From this, we can conclude that the issue lies in the implementation of down_aux and up_aux.

**Comment by YoucanBaby** [comment]
> I noticed that the original paper stated that down_aux/up_aux is a progressive orthogonal vector, but the code in this repository orthogonalized the auxiliary layer weights with down_dim/out_feature_dim. This may be the reason why losses cannot be converged. Can you take a look at it? Thanks!

The author mentioned a similar work on twitter. 
Hope this works for you.

[1] Z. Qiu et al., “Controlling Text-to-Image Diffusion by Orthogonal Finetuning,” arXiv preprint arXiv:2306.07280, 2023.

**Comment by KohakuBlueleaf** [comment]
I think the most possilbe reason is I actually got "different" aux when I rebuilt weight from pre_optimized weights
so it just didn't work

But I don't have time recently

If anyone have more experiments on this plz tell me or open PR

**Comment by 1585231086** [comment]
> 



> I think the most possilbe reason is I actually got "different" aux when I rebuilt weight from pre_optimized weights so it just didn't work
> 
> But I don't have time recently
> 
> If anyone have more experiments on this plz tell me or open PR


I have verified the code of reusing the aux by a random seed. It rightly uses the pre-training-stage aux. By the way, if using the trainable aux weights, I can indeed get a face image instead of something strange.

**Comment by KohakuBlueleaf** [comment]
> > 
> 
> > I think the most possilbe reason is I actually got "different" aux when I rebuilt weight from pre_optimized weights so it just didn't work
> > But I don't have time recently
> > If anyone have more experiments on this plz tell me or open PR
> 
> I have verified the code of reusing the aux by a random seed. It rightly uses the pre-training-stage aux. By the way, if using the trainable aux weights, I can indeed get a face image instead of something strange.

So the aux things may have something wrong
but I also have tried something similar before
"only train lora_up for lora" so lora_down are remain random

I have some idea, checking

**Comment by 1585231086** [comment]
> > > 
> > 
> > 
> > > I think the most possilbe reason is I actually got "different" aux when I rebuilt weight from pre_optimized weights so it just didn't work
> > > But I don't have time recently
> > > If anyone have more experiments on this plz tell me or open PR
> > 
> > 
> > I have verified the code of reusing the aux by a random seed. It rightly uses the pre-training-stage aux. By the way, if using the trainable aux weights, I can indeed get a face image instead of something strange.
> 
> So the aux things may have something wrong but I also have tried something similar before "only train lora_up for lora" so lora_down are remain random
> 
> I have some idea, checking


I found that the nn.init_orthogonal_()'s output will be different with the tensor's device even using the same random_seed. And another interesting thing is that if using xavier_normal initialization, the lilora weights can generate faces well(inspired by @chenxinhua ).

**Comment by chenxinhua** [comment]
I have retained orthogonal initialization for the weights, but made modifications to the initialization method of lilora and the implementation method of update_weight. By adjusting the learning strategy, I am now able to generate normal faces.

**Comment by 1585231086** [comment]
> I have retained orthogonal initialization for the weights, but made modifications to the initialization method of lilora and the implementation method of update_weight. By adjusting the learning strategy, I am now able to generate normal faces.

In my experiments, Only needed to change this repo's down_aux initial method to random initialization, the pre-optimization can work fine.

**Comment by KohakuBlueleaf** [comment]
> > I have retained orthogonal initialization for the weights, but made modifications to the initialization method of lilora and the implementation method of update_weight. By adjusting the learning strategy, I am now able to generate normal faces.
> 
> In my experiments, only needed to change this repo's down_aux initial method to random initialization, the pre-optimization can work fine.

is it possible to make pr for this?
We really need the fixes for this

**Comment by 1585231086** [comment]
> > > I have retained orthogonal initialization for the weights, but made modifications to the initialization method of lilora and the implementation method of update_weight. By adjusting the learning strategy, I am now able to generate normal faces.
> > 
> > 
> > In my experiments, only needed to change this repo's down_aux initial method to random initialization, the pre-optimization can work fine.
> 
> is it possible to make pr for this? We really need the fixes for this

Just change the lilora.py line nn.init.orthogonal_(down_aux,gain=1) to nn.init.xavier_normal_(down_aux,gain=1).

**Comment by SlZeroth** [comment]
nice work! is there any problem with train hypernetwork?

**Comment by 1585231086** [comment]
Hello, I found that the dataset definition may has something wrong. The index of image is needed to be sorted before the path_to_idx building:

```python
        self.instance_images_path = list(Path(instance_data_root).iterdir())
        self.path_to_idx = {
            str(path): idx 
            for idx, path in enumerate(sorted(self.instance_images_path))
        }
...
        img_path = self.instance_images_path[index % self.num_instance_images]
```

**Comment by chenxinhua** [comment]
1. I made some improvements to lilora, there is no need to perform orthogonal initialization for down_aux and up_aux every time when update_weight is executed.
2.  I also made some changes to train_preoptimized_liloras.py, the weights are saved directly for down_aux and up_aux matrices instead of aux_seed. 
3. Additionally, I added a script that can directly export lilora weights according to any given identity. During inference testing, it is now possible to generate normal faces.


Here are some generated faces based on the exported weights according to 2 given identities.

![PreOptHyperNet](https://github.com/KohakuBlueleaf/HyperKohaku/assets/17899959/bb7bf7be-1431-4724-9091-44936eaeb959)

**Comment by chenxinhua** [comment]
> can you share pretrained file? @chenxinhua

After further optimization, I will share the pretrained model.

**Comment by taemin6697** [comment]
I hope I can use it soon
@chenxinhua

**Comment by KohakuBlueleaf** [comment]
> > > > I have retained orthogonal initialization for the weights, but made modifications to the initialization method of lilora and the implementation method of update_weight. By adjusting the learning strategy, I am now able to generate normal faces.
> > > 
> > > 
> > > In my experiments, only needed to change this repo's down_aux initial method to random initialization, the pre-optimization can work fine.
> > 
> > 
> > is it possible to make pr for this? We really need the fixes for this
> 
> Just change the lilora.py line nn.init.orthogonal_(down_aux,gain=1) to nn.init.xavier_normal_(down_aux,gain=1).

done this

**Comment by KohakuBlueleaf** [comment]
> 1. I made some improvements to lilora, there is no need to perform orthogonal initialization for down_aux and up_aux every time when update_weight is executed.
> 2. I also made some changes to train_preoptimized_liloras.py, the weights are saved directly for down_aux and up_aux matrices instead of aux_seed.
> 3. Additionally, I added a script that can directly export lilora weights according to any given identity. During inference testing, it is now possible to generate normal faces.
> 
> Here are some generated faces based on the exported weights according to 2 given identities.
> 
> ![PreOptHyperNet](https://user-images.githubusercontent.com/17899959/266227209-bb7bf7be-1431-4724-9091-44936eaeb959.png)

your 1 and 2 is actually not a "improvements" which require extra vram/memory/storage for them
and lilora is for "you only need to store small amount of params"

for trianing process, cache aux may be ok but I think it should be optional
and for storage, I don't think store aux is a good idea

**Comment by 1585231086** [comment]
Does anybody check the train_hyperdreambooth.py? I found that the grads of weights are zero matrices in TransformerBlocks. Thanks.

**Comment by KohakuBlueleaf** [comment]
@1585231086 you are training with pre_optimized weights?

**Comment by 1585231086** [comment]
> @1585231086 you are training with pre_optimized weights?


Yes, I can ensure it. In fact, now I just used the pre_optimized weights as the gt weights instead of with the unet generation loss together.

**Comment by 1585231086** [comment]
> > @1585231086 you are training with pre_optimized weights?
> 
> Yes, I can ensure it. In fact, now I just used the pre_optimized weights as the gt weights instead of with the unet generation loss together.

My test is wrong. The grads of weights are normal.

**Comment by ChenYifu-s** [comment]
> 
the same problem

**Comment by Nnn-s** [comment]
> Does anybody check the train_hyperdreambooth.py? I found that the grads of weights are zero matrices in TransformerBlocks. Thanks.

the same problem, and my pre-optimized weight generates zero matrices as pre_opt_weights too

**Comment by taemin6697** [comment]
@chenxinhua
I can't reproduce that picture either.
Could you please share the code?

**Comment by taemin6697** [comment]
Is anyone working on this?

**Comment by kiyoon** [comment]
Previously, you had

```python
nn.init.orthogonal_(down_aux,gain=1)
nn.init.orthogonal_(up_aux,gain=1)
```

However, their dimensions are transposed, so I feel like the `orthogonal_` function for the `down_aux` has to be used with a transposed dimension.
In other words,

```python
# Note that I changed the down_aux dimension
down_aux = torch.empty(self.in_features, self.down_dim, dtype=weight.dtype, device=weight.device)
up_aux = torch.empty(self.out_features, self.up_dim, dtype=weight.dtype, device=weight.device)

# ...

nn.init.orthogonal_(down_aux,gain=1)
# And then change it back
down_aux = torch.t(down_aux)
nn.init.orthogonal_(up_aux,gain=1)
```
